File: middleware/js/internalAuthMiddleware.ts
import { Request, Response, NextFunction } from "express";
import crypto from "crypto";
import dotenv from "dotenv";
import path from "path";

// Load .env from project root
dotenv.config({ path: path.resolve(__dirname, "../../.env") });

/**
 * Constant-time string comparison to prevent timing attacks
 */
function timingSafeEqual(a: string, b: string): boolean {
  if (typeof a !== "string" || typeof b !== "string") {
    return false;
  }
  
  const bufferA = Buffer.from(a);
  const bufferB = Buffer.from(b);
  
  // If lengths differ, use dummy buffers to maintain constant time
  if (bufferA.length !== bufferB.length) {
    // Still compare to maintain constant time
    const dummyBuffer = Buffer.alloc(bufferA.length);
    crypto.timingSafeEqual(bufferA, dummyBuffer);
    return false;
  }
  
  return crypto.timingSafeEqual(bufferA, bufferB);
}

export function internalAuth(req: Request, res: Response, next: NextFunction) {
  const serviceSecret = req.headers["x-service-secret"];
  const isInternal = req.headers["x-internal-service"];
  const expectedSecret = process.env.INTERNAL_SERVICE_SECRET;

  // SECURITY: Use timing-safe comparison to prevent timing attacks
  if (
    !isInternal ||
    typeof serviceSecret !== "string" ||
    !expectedSecret ||
    !timingSafeEqual(serviceSecret, expectedSecret)
  ) {
    return res.status(403).json({ error: "Internal access only" });
  }
  next();
}

File: middleware/py/internal_auth_middleware.py
"""
Internal Authentication Middleware for FastAPI applications
Provides authentication for internal service-to-service communication
"""

from fastapi import Request, HTTPException, status
from starlette.middleware.base import BaseHTTPMiddleware
import os
import logging
import secrets

logger = logging.getLogger(__name__)


async def internal_auth(request: Request) -> None:
    """Dependency function for internal service authentication"""
    service_secret = request.headers.get("x-service-secret")
    is_internal = request.headers.get("x-internal-service")
    expected_secret = os.getenv("INTERNAL_SERVICE_SECRET")

    if not is_internal or not service_secret or not expected_secret or not secrets.compare_digest(
        service_secret, expected_secret
    ):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Internal access only",
        )


class InternalAuthMiddleware(BaseHTTPMiddleware):
    """Middleware for protecting routes with internal service authentication"""

    async def dispatch(self, request: Request, call_next):
        """Process request and validate internal service authentication"""
        # Check if route requires internal auth
        if hasattr(request.state, "require_internal_auth") and request.state.require_internal_auth:
            service_secret = request.headers.get("x-service-secret")
            is_internal = request.headers.get("x-internal-service")
            expected_secret = os.getenv("INTERNAL_SERVICE_SECRET")

            if not is_internal or not service_secret or not expected_secret or not secrets.compare_digest(
                service_secret, expected_secret
            ):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Internal access only",
                )

        response = await call_next(request)
        return response

